web/src/cypress/support/form.ts

```diff
@@ -1,14 +1,111 @@
-declare namespace Cypress {
-  interface Chainable<Subject> {
-    /** Update form fields with the given values. */
-    form: typeof form
+import { DateTime } from 'luxon'
+declare global {
+  namespace Cypress {
+    interface Chainable {
+      /** Update form fields with the given values. */
+      form: typeof form
+    }
   }
 }
 
+const clickArc = (pct: number) => (el: any) => {
+  const height = el.height() || 0
+  const radius = height / 2
+  const angle = (-1 + pct) * Math.PI * 2 - Math.PI / 2
+
+  const x = radius * Math.cos(angle) * 0.9 + radius
+  const y = radius * Math.sin(angle) * 0.9 + radius
+
+  return cy.wrap(el).click(x, y)
+}
+
+// materialClock will control a material time-picker from an input field
+function materialClock(time: string | DateTime) {
+  const dt = DateTime.isDateTime(time)
+    ? time
+    : DateTime.fromFormat(time, 'HH:mm')
+
+  let hour = dt.hour
+
+  const isAM = hour < 12
+  if (!isAM) hour -= 12
+
+  return cy
+    .contains('button', isAM ? 'AM' : 'PM')
+    .click() // select AM or PM
+
+    .get('[role=dialog][data-cy=picker-fallback] [role=menu]')
+    .parent()
+    .then(clickArc(hour / 12)) // select the hour
```

Comment:
It looks like we should be able to select the correct hour without doing any math to determine where the button is on the page. From what I'm seeing we can reference the picker once it's open and search for the span for each hour on the clock: `<span class="MuiTypography-root MuiPickersClockNumber-clockNumber MuiTypography-body1" style="transform: translate(55px, 19.6px);">1</span>`

E.g. maybe something like `cy.get('dialog picker-fallback menu span').contains(1).click()`

Reply:
It looks like you put a significant amount of thought into this though, so correct me if I'm wrong

Reply:
That could work for hours, but not minutes. Rather than having 2 code paths I just opted to use the same one so there's less to go wrong/debug.